### PR TITLE
[HOTT-477] Adds support for pulling out geographical_areas

### PR DESCRIPTION
**Pull request by willfish**
### Jira Link

https://transformuk.atlassian.net/browse/HOTT-477

### What

Adds support for pulling out geographical areas from the backend

### Why

We need this to pull out the European Union and it's members

### CODE REVIEW

File: spec/uktt_api_spec.rb [hunk 1]
```diff
@@ -444,4 +444,20 @@
       expect(response[:data].first[:attributes][:quota_order_number_id]).to eq(quota_search_params[:order_number])
     end
   end
+
+  it 'performs a retreive of countries' do
```
Comment (willfish) [inline spec/uktt_api_spec.rb]: These are not the tests I want to perform, btw. Ideally I'd do a fully mocked request verification but felt like I needed to at least test something. Coming later will be separately mocked specs for all of the resources catered by the uktt gem

File: lib/uktt.rb [hunk 2]
```diff
@@ -2,6 +2,7 @@
 require 'uktt/chapter'
 require 'uktt/commodity'
 require 'uktt/country'
+require 'uktt/geographical_area'
```
Comment (tavyy) [inline lib/uktt.rb:6]: 👍

File: lib/uktt/geographical_area.rb [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+      @response = nil
+    end
+
+    def retrieve
+      fetch "#{GEOGRAPHICAL_AREAS}"
+    end
+
+    def config=(new_opts = {})
+      merged_opts = Uktt.config.merge(new_opts)
+      Uktt.configure merged_opts
```
Comment (tavyy) [inline lib/uktt/geographical_area.rb]:
```suggestion
      Uktt.configure(merged_opts)
```

Comment (willfish) [reply lib/uktt/geographical_area.rb]: Probably we could find and replace this I reckon

File: lib/uktt/geographical_area.rb [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+      @config = Uktt.config
+    end
+
+    def find_in(arr)
+      return '@response is nil, run #retrieve first' unless @response
+  
+      response = @response.included.select do |obj|
+        arr.include?(obj.id)
+      end
+      response.length == 1 ? response.first : response
```
Comment (tavyy) [inline lib/uktt/geographical_area.rb]: So if response = [1], you return 1, but if it is [1,2,3,4] you return [1,2,3,4]... Hmmm, can we not just return an array all the time? E.g [1] or [1,2,3,4]

File: lib/uktt/geographical_area.rb [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    def config=(new_opts = {})
+      merged_opts = Uktt.config.merge(new_opts)
+      Uktt.configure merged_opts
+      @config = Uktt.config
+    end
+
+    def find_in(arr)
+      return '@response is nil, run #retrieve first' unless @response
+  
+      response = @response.included.select do |obj|
```
Comment (tavyy) [inline lib/uktt/geographical_area.rb]: You can use `response` here instead of `@response`

File: lib/uktt/geographical_area.rb [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+        arr.include?(obj.id)
+      end
+      response.length == 1 ? response.first : response
+    end
+
+    private
+
+    def fetch(resource)
+      @response = Uktt::Http.new(
+        @config[:host],
```
Comment (tavyy) [inline lib/uktt/geographical_area.rb]: You can use `config` here instead of `@config`

File: lib/uktt/geographical_area.rb (hunk 4 above)
Comment (willfish) [reply lib/uktt/geographical_area.rb]: I actually don't make use of this method, personally. I rely on the json api deserializer to give me an easy format

File: lib/uktt/geographical_area.rb [hunk 7]
```diff
@@ -0,0 +1,41 @@
+module Uktt
+  class GeographicalArea
+    attr_accessor :config, :response
+
+    def initialize(opts = {})
+      Uktt.configure(opts)
+      @config = Uktt.config
+      @response = nil
```
Comment (tavyy) [inline lib/uktt/geographical_area.rb]: I mean this will be nil unless assigned anyway?

File: lib/uktt/geographical_area.rb (hunk 4 above)
Comment (willfish) [reply lib/uktt/geographical_area.rb]: Probably we can remove it

File: spec/uktt_api_spec.rb (hunk 1 above)
Comment (tavyy) [inline spec/uktt_api_spec.rb]:
```suggestion
  it 'performs a retrieve of countries' do
```

File: spec/uktt_api_spec.rb [hunk 8]
```diff
@@ -444,4 +444,20 @@
...
+
+    expect_any_instance_of(Uktt::Http).to receive(:retrieve).with('geographical_areas/countries', 'jsonapi')
+
+    Uktt::Country.new.retrieve
+  end
+
+  it 'performs a retreive of geographical_areas' do
+    Uktt.configure(format: 'jsonapi', version: 'v2')
+
+    expect_any_instance_of(Uktt::Http).to receive(:retrieve).with('geographical_areas', 'jsonapi')
```
Comment (tavyy) [inline spec/uktt_api_spec.rb]: Hmm, I think that if you add a spy on Uktt::Http, you can just say

```ruby
expect(Uktt::Http).to have_received(:retrieve).with('geographical_areas', 'jsonapi')
```
?

File: lib/uktt/geographical_area.rb (hunk 7 above)
Comment (willfish) [reply lib/uktt/geographical_area.rb]: So I guess we can use these changes as part of a wider refactor. Probably we want a base class or a module we include

File: spec/uktt_api_spec.rb (hunk 8 above)
Comment (willfish) [reply spec/uktt_api_spec.rb]: Is an instance

**Review by tavyy — APPROVED** [review]
Amazing!

**Pull request merged**